Fixes #14295 - Initial version of Foreman Admin CLI

lzap opened the pull request:
The first cut. In the `test/fixture_tree` you can find files you can test
against. Here are diffs against unmodified versions of files deployed with
Satellite 6.2 beta installation (default options):

  hammer admin logging -ad

``` diff
diff -r ./etc/candlepin/candlepin.conf /tmp/test/etc/candlepin/candlepin.conf
43a44,45
> 
> log4j.logger.org.candlepin.servlet.filter=DEBUG
diff -r ./etc/dhcp/dhcpd.conf /tmp/test/etc/dhcp/dhcpd.conf
28c28
< log-facility local7;

---
> log-facility local6;
diff -r ./etc/foreman/settings.yaml /tmp/test/etc/foreman/settings.yaml
27c27
<   :level: info

---
>   :level: debug
diff -r ./etc/foreman-proxy/settings.yml /tmp/test/etc/foreman-proxy/settings.yml
55c55
< :log_level: ERROR

---
> :log_level: DEBUG
diff -r ./etc/mongod.conf /tmp/test/etc/mongod.conf
19a20,21
> 
> systemLog.verbosity=3
diff -r ./etc/pulp/server.conf /tmp/test/etc/pulp/server.conf
65c65
< # log_level:        The desired logging level. Options are: CRITICAL, ERROR, WARNING, INFO, DEBUG,

---
> # log_level: DEBUG
75c75
< log_level: INFO

---
> log_level: DEBUG
diff -r ./etc/puppet/puppet.conf /tmp/test/etc/puppet/puppet.conf
79a80,81
> 
> log_level=debug
diff -r ./etc/qpid/qpidd.conf /tmp/test/etc/qpid/qpidd.conf
30c30
< log-enable=error+

---
> log-enable=info+
Only in /tmp/test/etc: rsyslog.d
diff -r ./etc/sysconfig/virt-who /tmp/test/etc/sysconfig/virt-who
10c10
< VIRTWHO_DEBUG=0

---
> VIRTWHO_DEBUG=1
diff -r ./etc/xinetd.d/tftp /tmp/test/etc/xinetd.d/tftp
15c15
<         server_args     = -v -s /var/lib/tftpboot -m /etc/tftpd.map

---
>         server_args=-vvvv -s /var/lib/tftpboot -m /etc/tftpd.map
diff -r ./var/lib/pgsql/data/postgresql.conf /tmp/test/var/lib/pgsql/data/postgresql.conf
341c341
< #client_min_messages = notice         # values in order of decreasing detail:

---
> client_min_messages=debug2
352c352
< #log_min_messages = warning           # values in order of decreasing detail:

---
> log_min_messages=debug2
366c366
< #log_min_error_statement = error      # values in order of decreasing detail:

---
> log_min_error_statement=debug2
```

And after

  hammer admin logging -ad

``` diff
diff -r ./etc/candlepin/candlepin.conf /tmp/test/etc/candlepin/candlepin.conf
43a44,45
> 
> log4j.logger.org.candlepin.servlet.filter=WARN
diff -r ./etc/dhcp/dhcpd.conf /tmp/test/etc/dhcp/dhcpd.conf
28c28
< log-facility local7;

---
> log-facility local6;
diff -r ./etc/mongod.conf /tmp/test/etc/mongod.conf
19a20,21
> 
> systemLog.verbosity=0
diff -r ./etc/pulp/server.conf /tmp/test/etc/pulp/server.conf
65c65
< # log_level:        The desired logging level. Options are: CRITICAL, ERROR, WARNING, INFO, DEBUG,

---
> # log_level: INFO
diff -r ./etc/puppet/puppet.conf /tmp/test/etc/puppet/puppet.conf
79a80,81
> 
> log_level=notice
Only in /tmp/test/etc: rsyslog.d
diff -r ./etc/xinetd.d/tftp /tmp/test/etc/xinetd.d/tftp
15c15
<         server_args     = -v -s /var/lib/tftpboot -m /etc/tftpd.map

---
>         server_args=-v -s /var/lib/tftpboot -m /etc/tftpd.map
diff -r ./var/lib/pgsql/data/postgresql.conf /tmp/test/var/lib/pgsql/data/postgresql.conf
341c341
< #client_min_messages = notice         # values in order of decreasing detail:

---
> client_min_messages=notice
352c352
< #log_min_messages = warning           # values in order of decreasing detail:

---
> log_min_messages=warning
366c366
< #log_min_error_statement = error      # values in order of decreasing detail:

---
> log_min_error_statement=error
```

May I ask @tstrachota for review? I will soon provide you with RPM packages so
you can test on real Satellite instance. Until then, please do code review.
Feel free to test against fixtures too.

lzap committed:
Fixes #14295 - Initial version of Foreman Admin CLI

tstrachota commented:
I'll take a look, @lzap 
First thing I noticed is you're missing translations everywhere. Was that intentional? You can find an example how to set it up for a plugin here: https://github.com/theforeman/hammer-cli-foreman/blob/master/lib/hammer_cli_foreman/i18n.rb

tstrachota commented:
It works nice, @lzap. I like the configurable actions in the config files. I found a few issues during my test:

1) It raises a `NoMethodError` exception when the config files are missing. It's quite confusing. It would be better to let user know what happened.

```
NoMethodError (undefined method `[]' for nil:NilClass):
    /root/hammer/hammer-cli-foreman-admin/lib/hammer_cli_foreman_admin/logging_command.rb:111:in `execute'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
    /root/hammer/hammer-cli/lib/hammer_cli/abstract.rb:22:in `run'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
    /root/hammer/hammer-cli/lib/hammer_cli/abstract.rb:22:in `run'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
    /root/hammer/hammer-cli/lib/hammer_cli/abstract.rb:22:in `run'
    /usr/local/rvm/gems/ruby-2.0.0-p643/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run'
    ./bin/hammer:125:in `<main>'
```

2) Both normal action and dryrun prints only to the log. I wasn't sure whether the command did anything. Could you print some message to stdout?

3) The command failed silently when the config file wasn't writable (`$?` was 0 too).

4) It fails silently when you do `hammer -d admin logging --no-backup  -p --components UNKNOWN_COMPONENT` I'd expect an error message in such case.

tstrachota commented on lib/hammer_cli_foreman_admin/logging_command.rb:
```diff
@@ -0,0 +1,127 @@
+    end
+
+    def configure_component(component, level)
+      name = component[:name]
```
Could you add something like `check_options` here too, please? When I tried to edit the settings and misconfigured them, I usually end up with `NoMethodError (undefined method`[]' for nil:NilClass)`

lzap replied:
Isn't this solved by checking from list of available components?

tstrachota replied:
I don't think so. I meant the cases when the settings file is valid yaml, but expected keys are missing. For example when you add empty bullet to the list of components. But it's a minor thing, not important.

tstrachota commented on lib/hammer_cli_foreman_admin/logging_command.rb:
```diff
@@ -0,0 +1,127 @@
+    def list_components
+      HammerCLI::Settings.get(:admin)[:logging][:component].each do |component|
+        print_message(component[:name])
+      end
```
Have you considered printing the values in a table? I don't care about formatting but it would enable for printing the list in csv and json (`--output=json`).

The code would be something like:

``` ruby
output_definition = HammerCLI::Output::Definition.new
output_definition.fields << Fields::Field.new(:label => _('Component'), :path => ["name"])
output = HammerCLI::Output::Output.new(context, :default_adapter => :table)
output.print_collection(output_definition, HammerCLI::Settings.get(:admin)[:logging][:component])
```

lzap replied:
This renders as

```
----------
COMPONENT 
----------
postgresql
rails     
proxy     
puppet    
dhcpd     
named     
tftp      
mongod    
qpidd     
tomcat    
pulp      
virt-who  
----------
----------
COMPONENT 
----------
postgresql
rails     
proxy     
puppet    
dhcpd     
named     
tftp      
mongod    
qpidd     
tomcat    
pulp      
virt-who  
----------
----------
COMPONENT 
----------
postgresql
rails     
proxy     
puppet    
dhcpd     
named     
tftp      
mongod    
qpidd     
tomcat    
pulp      
virt-who  
----------
```

Can you give me hint how to do that correctly?

lzap replied:
I take it back, fixed:

```
---------------|------------------------------------
COMPONENT NAME | AUTO-DETECTED BY EXISTENCE OF      
---------------|------------------------------------
postgresql     | /var/lib/pgsql/data/postgresql.conf
rails          | /etc/foreman/settings.yaml         
proxy          | /etc/foreman-proxy/settings.yml    
puppet         | /etc/puppet/puppet.conf            
dhcpd          | /etc/dhcp/dhcpd.conf               
named          | /etc/named.conf                    
tftp           | /etc/xinetd.d/tftp                 
mongod         | /etc/mongod.conf                   
qpidd          | /etc/qpid/qpidd.conf               
tomcat         | /etc/candlepin/candlepin.conf      
pulp           | /etc/pulp/server.conf              
virt-who       | /etc/sysconfig/virt-who            
---------------|------------------------------------
```

lzap commented:
Added i18n in a separate commit (will squash later on).

lzap commented:
1) Fixed missing configuration problem.

2) I want to follow traditional UNIX behavior - until verbose mode is provided, do not print anything. I haven't implemented verbose mode at all, looks like it's a bit messy right now in Hammer Core.

3) I get an error, note that not all config files are affected, for example named.conf is not touched, it is only used for detection if bind was installed. Then a command is executed to increase verbosity, in this case it's "rndc querylog xyz". If that fails, then we do not fail hard but only issue an error: Command 'rndc querylog off' failed: No such file or directory - rndc.

When I set readonly on other file, then it fails correctly:

```
Error: Permission denied - /home/lzap/work/hammer-cli-foreman-admin/test/fixture_tree/etc/mongod.conf
```

4) Added extra check for that

lzap commented:
Pushed (backup of the original idea is at http://paste.fedoraproject.org/357688/61143481 just for the record)

``` diff
diff --git a/lib/hammer_cli_foreman_admin/logging_command.rb b/lib/hammer_cli_foreman_admin/logging_command.rb
index 71643ec..6449281 100644
--- a/lib/hammer_cli_foreman_admin/logging_command.rb
+++ b/lib/hammer_cli_foreman_admin/logging_command.rb
@@ -106,21 +106,23 @@ module HammerCLIForemanAdmin
       end
     end

-    def list_components
-      HammerCLI::Settings.get(:admin)[:logging][:component].each do |component|
-        print_message(component[:name])
-      end
+    def available_components
+      HammerCLI::Settings.get(:admin)[:logging][:component]
     end

-    # cp -r /home/lzap/work/hammer-cli-foreman-admin/test /tmp
     def execute
-      configuration = HammerCLI::Settings.get(:admin)[:logging][:component]
+      configuration = HammerCLI::Settings.get(:admin)[:logging][:component] rescue raise("Missing logging YAML definitions (foreman_admin_logging_*.yml)")
       if option_list?
-        list_components
+        output_definition = HammerCLI::Output::Definition.new
+        output_definition.fields << Fields::Field.new(:label => _('Component name'), :path => ["name"])
+        output_definition.fields << Fields::Field.new(:label => _('Auto-detected by existence of'), :path => ["file"])
+        output = HammerCLI::Output::Output.new(context, :default_adapter => :table)
+        output.print_collection(output_definition, HammerCLI::Settings.get(:admin)[:logging][:component])
       else
         if option_all?
           components = configuration
         else
+          raise("Unknown component provided, use --list to find them") unless option_components.all? { |c| available_components.include? c }
           components = configuration.select{ |x| option_components.include? x[:name] }
         end
         components.each { |component| configure_component(component, new_level) }
```

If you dont have any other objections, I would like to go ahead with 0.0.1 release for some initial testing.

tstrachota commented on lib/hammer_cli_foreman_admin/logging_command.rb:
```diff
@@ -0,0 +1,135 @@
+      if raise_on_error
+        raise "Command '#{cmd}' failed: #{e}"
+      else
+        print_message "Command '#{cmd}' failed: #{e}"
```
Since this is exposed to users, I believe the message should be translated.

tstrachota commented:
One note regarding translations. Please consider using gettext for exceptions that are exposed to users. Otherwise :+1: to merge.

lzap commented:
Done, thanks!
